### ISSUE TITLE
Component wont be updated in production build.

### ISSUE THREAD

**Issue by Albert-Gao** [mobxjs/mobx-react-lite#14]
**My packages:**
```json
    "mobx": "^4.6.0",
    "mobx-react-lite": "^0.2.0",
    "normalize.css": "^8.0.1",
    "react": "^16.7.0-alpha.2",
    "react-dom": "^16.7.0-alpha.2",
    "react-pose": "^4.0.1",
    "react-router-dom": "^4.3.1",
    "react-scripts": "^2.1.1",
    "react-typing-animation": "^1.4.0",
    "rebass": "^3.0.0-11",
    "styled-components": "^4.1.1",
    "styled-system": "^3.1.11",
    "styled-tools": "^1.6.0",
    "typescript": "^3.1.6"
```

**What's the problem:**
1. I have an `AppStore.tsx`:
```typescript
export default class AppStore {
  @observable
  steps: ReactElementArray = [];

  @action.bound
  nextStep() {}
}
```

The logic is, there will be a component array. After each step finishes, it will call `nextStep()`, then a new step(a component) will be pushed into `this.steps`.

2. And it will trigger render of this `<ChatScreen>`:
```typescript
import React, { useContext } from 'react';
import { Flex } from 'rebass';
import { RouteComponentProps } from 'react-router-dom';
import { observer } from 'mobx-react-lite';
import { PoseGroup } from 'react-pose';
import Header from '../../components/Header/Header';
import { Box } from '../../components/buildingBlocks';
import { AppStoreContext } from '../../store/store';

const ChatScreen: React.FC<
  RouteComponentProps
> = () => {
  const { steps } = useContext(AppStoreContext);

  return (
    <Flex width={1} flexDirection="column">
      <Header title="i am empty title" />
      <Box>
        <PoseGroup animateOnMount>{steps}</PoseGroup>
      </Box>
    </Flex>
  );
};

export default observer(ChatScreen);
```

3. Everything works fine in the development build but not the production build. The production build will execute step 1, but after a new step been added to `this.step`(I checked, the length increased). It won't trigger re-render of ChatScreen.

4. Basically, the component of second step won't be rendered at all.
```typescript
const LocationDropDown: React.FC<IBaseStepType> = ({
  callback,
}) => {
  console.log('i am LocationDropDown');
  return (
    <FlexCard boxShadow={2}>
      <ProvincesDropDown />
      <CitiesDropDown callback={callback} />
    </FlexCard>
  );
};

export default observer(LocationDropDown);
``` 
Because the `console.log` won't be shown in the console.

Any idea why?

It's a private repo. Invitation can be sent.

I can't go back to the no-hook and `inject` age for this green field project, no.........

**Comment by mweststrate** [comment]
The chatscreen shouldn't re-render upon change of `steps`, because there is no reason for it to render (the output wouldn't modify), as you are not using anything in `steps`. If you change your rendering to for example `<PoseGroup animateOnMount>{steps.length}</PoseGroup>` you should see a render when you add an item, but not when you update etc. 

Remember: MobX tracks what you _use_. `steps` is just a pointer to an array, that pointer itself never changes, so no update will ever be triggered. 

For more in-depth: https://mobx.js.org/best/react.html

That being said, it is unclear to me why it _does_ re-render in dev mode :). If you have any future questions, make sure to set up small code sandbox demonstrating the problem

**Comment by mweststrate** [comment]
N.B. Hooks are not official yet, so make sure you are not using this in a project which you actually need to release anywhere soonish :)

**Comment by Albert-Gao** [comment]
Thanks for the advice. It's a personal project. :)

That is interesting. In `<ChatScreen>`,  I just updated:
```javascript
<PoseGroup animateOnMount>{steps}</PoseGroup>
``` 
to
```javascript
<PoseGroup animateOnMount>{steps}</PoseGroup>
{steps.length}
```

Now everything works in production build as well.........

Won't work if remove that `{steps.length}`...

**Comment by Albert-Gao** [comment]
So create a repo with minimal code https://github.com/Albert-Gao/notRender , I think Michael is right.
It shouldn't render.

If I don't wrap `{steps}` with that `<PoseGroup>`, even with that `{steps.length}`, nothing happens.
So, seems `{steps.length}` triggers the re-render.

Michael mentioned something:
> The chatscreen shouldn't re-render upon change of steps, because there is no reason for it to render (the output wouldn't modify), as you are not using anything in steps.

But since `steps` is an observable, when it's been updated, shouldn't component like `<ChatScreen>` which subscribe to it using `observer` auto-updates itself?

**Comment by Albert-Gao** [comment]
But strangely, in dev mode, even without that `steps.length`, everything works as expected.

**Comment by FredyC** [comment]
So I've run your reproduction and indeed it does not work in a production build. However, observing over the reference to `steps` is not right as @mweststrate described. I've prepared an even more minimalistic example that should be the same in its essence.

https://codesandbox.io/s/rrl7pvj6lp

If you click on `Add step` you see that only the second list is actually re-rendered as it should be because `Array.from` does the iteration over `steps` elements. I've also tried downloading that sandbox and making a production build - exactly the same result.

You can do the same trick in your code. I've tried it and works well in a production too. I am not sure what's wrong in your code/setup really.

**Comment by FredyC** [comment]
Whoa, this is really weird. I did a slight modification to that codesandbox repro and instead of passing steps in a separate prop, I've moved it to children. Suddenly, both cases work. And if I try the production build, it's broken - does not work without Array.from.

https://codesandbox.io/s/6w79o4p03k

@mweststrate Any idea why would children be observed and regular prop not? Looks like some weird bug indeed. But be in MobX itself since this package is doing very little.

**Comment by vensa-albertgao** [comment]
`Array.from` will generate a new array, so I think that's why it works, so each render will yield a new array, it will render, a `key` attribute here should prevent unnecessary re-render.

I think with all the inputs, I understand why the reference won't work now, so underneath, of course, mobx is not like redux, every time, the `observable` in mobx store updates, it's not a new object? Which means its address in memory won't change, so the reference won't change as well. And if I observe the reference, of course, it shouldn't work. Because `not using anything in `steps`!`

Thanks @mweststrate and @FredyC !!!

What about a new PR to add this to the doc? They are all very straightforward but when people get used to mobx's power, they might miss something.

`Array.from` is a pretty good solution! Very tidy!

I'm thinking of moving the generating `steps` problems to a component rather than in the store, and this component would observe either `steps.length` or another observable in store to indicate whether new steps have been pushed or not. That should perfectly work.

But yeah, why the children works... don't know.

**Comment by FredyC** [comment]
I wouldn't recommend `Array.from` as some permanent solution, it's merely a hack in my opinion. There is definitely a bug in different behavior between children and regular prop. I hope @mweststrate can bring some light into this, I am totally clueless.

**Comment by mweststrate** [comment]
I think calling a property `children` can be troublesome as it is treated specially by react, so I can image React always clones properties called children, or treats them specially, causing this to break.

The conversation is a bit hard to follow, but it is still unclear to me why Array.from would be needed to render arrays?

**Comment by mweststrate** [comment]
So  you are passing the array to a component that isn't an observer?
Otherwise that shouldn't be needed. Note that `.slice()` is easier in
general and does the same :)

Op ma 19 nov. 2018 om 12:25 schreef AlbertGao <notifications@github.com>:

**Comment by FredyC** [comment]
@mweststrate Any chance you can have a look at those reproductions? I am still convinced this goes beyond this package.

**Comment by mweststrate** [comment]
@FredyC sorry, no not yet! Will catch up with the open issues on this repo soon :)

**Comment by bogdandjukic** [comment]
Watching this one.
I got the same problem. Tried what you guys said works, and replicated completely the same.
Works only if passed as `{children}` or if with `Array.from`.
Definitely a strange one.

**Comment by bogdandjukic** [comment]
Also I noticed that if you import store as `useContext` directly into the component it also works, but then the component is tied to that specific array and it's not a 'dumb' component anymore.

**Comment by urugator** [comment]
Imo react simply checks `children.length` or something on devel, causing the subscription.
Similar problem occurs when prop types like `arrayOf(...)` are defined.

The behavior on production is as expected, the `steps` aren't accessed inside `observer/Observer` therefore there is no subscription. To make it obvious, here is a simplification of the sandobox:
```javascript
// steps untouched inside Observer
<Observer>{() => <StepsDisplayProp steps={steps} />}</Observer>
<Observer>{() => <StepsDisplayChildren>{steps}</StepsDisplayChildren>}</Observer>
// steps touched by Array.from inside Observer 
<Observer>{() => <StepsDisplayProp steps={Array.from(steps)} />}</Observer>
<Observer>{() => <StepsDisplayChildren>{Array.from(steps)}</StepsDisplayChildren>}</Observer>

// Note that what happens in <StepsDisplayProp/Children> is absolutely irrelevant
// as they're not observers, they could just return Date.now() and update the same way
```

@Albert-Gao
It is the `PoseGroup` which depends on `steps` content, not `ChatScreen`, therefore `PoseGroup` must be an `observer`. If it can't (eg it's 3rd party), you have to subscribe in `ChatScreen` (which must be `observer`) manually with `steps.length`/`toJS(steps)`/`Array.from(steps)`/anything which touches the array's content:
https://github.com/mobxjs/mobx-react/issues/484#issuecomment-391685354
The fact that `{children}` works on devel is probably just a coincidence as pointed above.

**Comment by FredyC** [comment]
Ok, I think @urugator got the gist of the issue and it makes sense. Closing for now.
